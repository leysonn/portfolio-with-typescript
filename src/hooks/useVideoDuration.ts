import { useMemo, useState } from 'react';

/**
 * Custom hook to get the video duration and a function to set the metadata.
 *
 * @return {[string, (event: SyntheticEvent<HTMLVideoElement>) => void]} An array containing the formatted video duration and a function to set the metadata.
 */
function useVideoDuration(): [string, (duration: number) => void] {
    const [duration, setDuration] = useState(0);

    const formattedDuration = useMemo(() => {
        if (duration < 0 || isNaN(duration)) {
            return '0:00';
        }

        const minutes = Math.floor(duration / 60);
        const remainingSeconds = Math.floor(duration % 60);
        const formattedSeconds = remainingSeconds < 10 ? '0' + remainingSeconds : remainingSeconds;

        return minutes + ':' + formattedSeconds;
    }, [duration]);

    function setMetadata(duration: number): void {
        setDuration(duration);
    }

    return [formattedDuration, setMetadata];
}

export { useVideoDuration };
