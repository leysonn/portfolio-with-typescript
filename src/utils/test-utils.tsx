import { Outlet, RouterProvider, createRootRoute, createRoute, createRouter } from '@tanstack/react-router';
import { cleanup, render } from '@testing-library/react';
import { afterEach } from 'vitest';

afterEach(() => {
    cleanup();
});

function createTestRouter(component: () => React.ReactElement) {
    const rootRoute = createRootRoute({
        component: Outlet,
    });

    const componentRoute = createRoute({
        getParentRoute: () => rootRoute,
        path: '/',
        component,
    });

    const router = createRouter({
        routeTree: rootRoute.addChildren([componentRoute]),
    });

    return router;
}

function renderWithRouter(ui: () => React.ReactElement, options = {}) {
    const router = createTestRouter(ui);
    return render(<RouterProvider router={router} />, options);
}

// eslint-disable-next-line react-refresh/only-export-components
export * from '@testing-library/react';
export { renderWithRouter };
