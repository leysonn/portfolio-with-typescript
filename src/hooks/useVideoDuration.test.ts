import { act, renderHook, waitFor } from '../utils/test-utils';
import { useVideoDuration } from './useVideoDuration';

describe('useVideoDuration', () => {
    it('call without duration', () => {
        const { result } = renderHook(() => useVideoDuration());
        expect(result.current[0]).toBe('0:00');
    });

    it('call with correct duration', async () => {
        const { result } = renderHook(useVideoDuration);
        act(() => result.current[1](100));
        await waitFor(() => result.current[0] !== '0:00');
        expect(result.current[0]).toMatchSnapshot();
    });

    it('call with boundary duration', async () => {
        const { result, rerender } = renderHook(() => useVideoDuration());
        act(() => result.current[1](1.9999));
        await waitFor(() => expect(result.current[0]).not.toBe('0:00'));
        expect(result.current[0]).toMatchSnapshot();

        rerender();
        act(() => result.current[1](100000));
        await waitFor(() => expect(result.current[0]).not.toBe('0:00'));
        expect(result.current[0]).toMatchSnapshot();
    });
});
