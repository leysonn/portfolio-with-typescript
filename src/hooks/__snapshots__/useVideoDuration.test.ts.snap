// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useVideoDuration > call with boundary duration 1`] = `"0:01"`;

exports[`useVideoDuration > call with boundary duration 2`] = `"1666:40"`;

exports[`useVideoDuration > call with correct duration 1`] = `"1:40"`;
